// logger-helper/index.js
var log4js = require("log4js");
const _ = require('lodash');
const defaultOptions =
	{
		appenders: {
			console: {
				type: 'console',
				layout: {
					type: 'pattern',
					pattern: '%[%d{yyyy-MM-dd hh:mm:ss:SSS} pid:%x{pid} %c [%p]: %] %m',
					tokens: {
						pid: function () {
							return process.pid;
						}
					}
				}
			},
			/*dateLog: {
				type: 'dateFile', filename: "logs/web",
				pattern: "-yyyy-MM-dd.log", alwaysIncludePattern: true,
				pollInterval: 1,
				layout: {
					type: 'pattern',
					pattern: '[%d{yyyy-MM-dd hh:mm:ss:SSS}] [%p]:%m'
				}
			},*/
			file:{
				type: "file",
				filename : "logs/app.log",
				maxLogSize : 100, // 字节B 204800
				backups : 3,
				encoding: 'utf-8',
				// compress: true,
				layout: {
					type: 'pattern',
					pattern: '[%d{yyyy-MM-dd hh:mm:ss:SSS}] pid:%x{pid} [%p]:%m', //category:%c
					tokens: {
						pid: function () {
							return process.pid;
						}
					}
				}
			}
		},
		categories: {
			default:{appenders: ['console'], level: 'debug'},
			// dateLog: {appenders: ['dateLog'], level: 'debug'},
			file:{appenders: ['file'], level: 'debug'}
		},
		pm2: true, // pm2 install pm2-intercom
		pm2InstanceVar: 'INSTANCE_ID'
	};
var LOGGER = function (options,name) {
	var options = _.merge({}, defaultOptions, options);
	log4js.configure(options); // log_config
	const logger = log4js.getLogger();
	console.log = logger.info.bind(logger);
	var fileLogger = {};
	fileLogger.logger = log4js.getLogger(name);
	fileLogger.use=function () { // for koa framework
		return function (ctx, next) {
			ctx.logger = log4js.getLogger(name);
			return next();
		};
	};
	fileLogger.location=function(){
		var Location = "";
		try {
			throw new Error();
		} catch (e) {
			// console.log("Stack:" + e.stack);
			Location= "Location: " + e.stack.replace(/Error\n/).split(/\n/)[1].replace(/^\s+|\s+$/, "");
		}
		// console.log(Location);
		return Location;
	};
	return fileLogger;
};
module.exports = LOGGER;
